File: 0688_Knight_Probability_in_Chessboard/solution.java
package cn.huang.leetcode;

/*
688. Knight Probability in Chessboard

On an NxN chessboard, a knight starts at the r-th row and c-th column and attempts to make exactly K moves.
The rows and columns are 0 indexed, so the top-left square is (0, 0), and the bottom-right square is (N-1, N-1).

A chess knight has 8 possible moves it can make, as illustrated below. Each move is two squares in a cardinal direction,
then one square in an orthogonal direction.


Each time the knight is to move, it chooses one of eight possible moves uniformly at random (even if the piece would go
off the chessboard) and moves there.

The knight continues moving until it has made exactly K moves or has moved off the chessboard. Return the probability
that the knight remains on the board after it has stopped moving.


Example:

Input: 3, 2, 0, 0
Output: 0.0625
Explanation: There are two moves (to (1,2), (2,1)) that will keep the knight on the board.
From each of those positions, there are also two moves that will keep the knight on the board.
The total probability the knight stays on the board is 0.0625.


Note:

    N will be between 1 and 25.
    K will be between 0 and 100.
    The knight always initially starts on the board.

 */
public class LC_0688_KnightProbabilityinChessboard {
    /*
    https://www.youtube.com/watch?v=OrS7PaJ-5ck
     */
    public double knightProbability(int N, int K, int r, int c) {
        int[][] dirs = new int[][]{{1, 2}, {1, -2}, {-1, 2}, {-1, -2}, {2, 1}, {2, -1}, {-2, 1}, {-2, -1}};
        double[][][] dp = new double[K + 1][N][N];

        dp[0][r][c] = 1;
        for (int k = 1; k <= K; k++) {
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    for (int[] dir : dirs) {
                        int next_i = i + dir[0], next_j = j + dir[1];
                        if (next_i < 0 || next_j < 0 || next_i >= N || next_j >= N)
                            continue;
                        dp[k][next_i][next_j] += dp[k - 1][i][j];
                    }
                }
            }
        }

        double cnt = 0;
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                cnt += dp[K][i][j];
        return cnt / Math.pow(8, K);
    }
}

File: 0688_Knight_Probability_in_Chessboard/solution.cpp
/*
688. Knight Probability in Chessboard

Level: Medium

https://leetcode.com/problems/knight-probability-in-chessboard
*/
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

class Solution
{
public:
  double knightProbability(int N, int K, int r, int c)
  {
    vector<vector<double>> dp0(N, vector<double>(N, 0.0));
    dp0[r][c] = 1.0;
    int dirs[8][2] = {{1, 2}, {-1, -2}, {1, -2}, {-1, 2}, {2, 1}, {-2, -1}, {2, -1}, {-2, 1}};
    for (int k = 0; k < K; ++k)
    {
      vector<vector<double>> dp1(N, vector<double>(N, 0.0));
      for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
          for (int m = 0; m < 8; ++m)
          {
            int x = j + dirs[m][0];
            int y = i + dirs[m][1];
            if (x < 0 || y < 0 || x >= N || y >= N)
              continue;
            dp1[i][j] += dp0[y][x];
          }
      std::swap(dp0, dp1);
    }

    double total = 0;
    for (int i = 0; i < N; ++i)
      for (int j = 0; j < N; ++j)
        total += dp0[i][j];

    return total / pow(8, K);
  }
};